src/main/java/core/stack/MateStack.java

```diff
@@ -1,19 +1,57 @@
 package core.stack;
 
+import java.util.EmptyStackException;
+
 public class MateStack<T> {
-    public void push(T value) {
+    private static final int INITIAL_SIZE = 10;
+    private int currentSize = INITIAL_SIZE;
+    private int size = 0;
+    private T[] table = (T[]) new Object[currentSize];
 
+    public void push(T value) {
+        if (size >= currentSize) {
+            table = grow();
+        }
+        table[size] = value;
+        size++;
     }
 
     public T peek() {
-        return null;
+        if (size == 0) {
+            throw new EmptyStackException();
+        }
+        return table[size - 1];
     }
 
     public T pop() {
-        return null;
+        if (size == 0) {
+            throw new EmptyStackException();
+        }
+        T value = peek();
+        table[size - 1] = null;
+        size--;
+        return value;
     }
 
     public int size() {
-        return 0;
+        return size;
+    }
+
+    private T[] grow() {
+        double newLength = table.length * 1.5;
+        currentSize = (int) newLength;
+        T[] newTable = (T[]) new Object[currentSize];
+        for (int i = 0; i < table.length; i++) {
+            newTable[i] = table[i];
+        }
```

Comment:
You can use `System.arraycopy` to copy the elements from the old array to the new one. This is likely to be more efficient than a manual loop, as `System.arraycopy` is implemented as a native method.

Reply:
ok, done